File: sdk/javascript/src/platform-client.ts
import { requestJson, type HttpClientOptions } from "./http";
import { DatabaseContext } from "./database/database-context";
import { DtorchValidationError } from "./errors";
import { EltRealtimeClient } from "./realtime-client";
import { EltRuntimeClient } from "./runtime-client";
import type {
  WorkspaceDatabaseListResponse,
  WorkspaceDatabaseSqlResponse,
  WorkspaceDatabaseTableDataResponse,
  WorkspaceDatabaseTableDetail,
  WorkspaceDatabaseTableListResponse,
} from "./types/database";

export type WorkspaceStorageStatus = {
  workspace_id: number;
  status: string;
  mode?: string | null;
  bucket?: string | null;
  prefix?: string | null;
  endpoint_url?: string | null;
  plan?: string | null;
  s3_uri?: string | null;
  provisioned_at?: string | null;
  error?: string | null;
  can_provision: boolean;
};

export type StorageObjectItem = {
  key: string;
  name: string;
  type: string;
  size: string;
  size_bytes?: number | null;
  modified?: string | null;
};

export type EltPlatformClientOptions = {
  baseUrl: string;
  projectKey: string;
  projectSecret: string;
  workspaceId: number;
  databaseId?: number;
  timeoutMs?: number;
};

export class EltPlatformClient {
  private http: HttpClientOptions;
  readonly storage: {
    getStatus: () => Promise<WorkspaceStorageStatus>;
    listObjects: (prefix?: string) => Promise<{ items: StorageObjectItem[]; prefix: string }>;
    uploadObject: (
      file: Blob,
      options?: { key?: string; fileName?: string },
    ) => Promise<{ key: string; name: string }>;
    deleteObject: (key: string) => Promise<void>;
  };
  readonly runtime: EltRuntimeClient;
  readonly realtime: EltRealtimeClient;
  readonly workspaceId: number;
  readonly defaultDatabaseId?: number;

  constructor(options: EltPlatformClientOptions) {
    this.workspaceId = options.workspaceId;
    this.defaultDatabaseId = options.databaseId;
    this.http = {
      baseUrl: options.baseUrl,
      auth: {
        type: "project",
        projectKey: options.projectKey,
        projectSecret: options.projectSecret,
      },
      timeoutMs: options.timeoutMs,
    };
    this.runtime = new EltRuntimeClient({
      baseUrl: options.baseUrl,
      projectKey: options.projectKey,
      projectSecret: options.projectSecret,
      workspaceId: options.workspaceId,
      timeoutMs: options.timeoutMs,
    });
    this.realtime = new EltRealtimeClient({
      baseUrl: options.baseUrl,
      projectKey: options.projectKey,
      projectSecret: options.projectSecret,
      workspaceId: options.workspaceId,
      timeoutMs: options.timeoutMs,
    });
    this.storage = {
      getStatus: () => this.getStorageStatus(),
      listObjects: (prefix) => this.listStorageObjects(prefix),
      uploadObject: (file, uploadOptions) => this.uploadStorageObject(file, uploadOptions),
      deleteObject: (key) => this.deleteStorageObject(key),
    };
  }

  database(databaseId: number): DatabaseContext {
    return new DatabaseContext(this, databaseId);
  }

  /** Default database namespace configured with `databaseId`. */
  get db(): DatabaseContext {
    if (this.defaultDatabaseId === undefined) {
      throw new DtorchValidationError(
        "No default database configured; pass databaseId or call database(id)",
      );
    }
    return this.database(this.defaultDatabaseId);
  }

  /** Validate project credentials against the linked workspace. */
  async validate() {
    const [validation, databases] = await Promise.all([
      requestJson<{ ok: true; workspace_id: number; scopes: string[] }>(
        this.http,
        "GET",
        `/api/v1/workspaces/${this.workspaceId}/project/validate`,
      ),
      this.listDatabases(),
    ]);
    return {
      ok: validation.ok,
      workspaceId: validation.workspace_id,
      scopes: validation.scopes,
      databases: databases.databases,
    };
  }

  listDatabases() {
    return requestJson<WorkspaceDatabaseListResponse>(
      this.http,
      "GET",
      `/api/v1/workspaces/${this.workspaceId}/databases`,
    );
  }

  listTables(databaseId: number) {
    return requestJson<WorkspaceDatabaseTableListResponse>(
      this.http,
      "GET",
      `/api/v1/workspaces/${this.workspaceId}/databases/${databaseId}/tables`,
    );
  }

  getTableDetail(databaseId: number, tableName: string) {
    return requestJson<WorkspaceDatabaseTableDetail>(
      this.http,
      "GET",
      `/api/v1/workspaces/${this.workspaceId}/databases/${databaseId}/tables/${encodeURIComponent(tableName)}`,
    );
  }

  getTableData(
    databaseId: number,
    tableName: string,
    opts?: { limit?: number; offset?: number },
  ) {
    const params = new URLSearchParams();
    if (opts?.limit !== undefined) params.set("limit", String(opts.limit));
    if (opts?.offset !== undefined) params.set("offset", String(opts.offset));
    const q = params.toString() ? `?${params.toString()}` : "";
    return requestJson<WorkspaceDatabaseTableDataResponse>(
      this.http,
      "GET",
      `/api/v1/workspaces/${this.workspaceId}/databases/${databaseId}/tables/${encodeURIComponent(tableName)}/data${q}`,
    );
  }

  executeSql(databaseId: number, sql: string) {
    return requestJson<WorkspaceDatabaseSqlResponse>(
      this.http,
      "POST",
      `/api/v1/workspaces/${this.workspaceId}/databases/${databaseId}/sql`,
      { sql },
    );
  }

  getStorageStatus() {
    return requestJson<WorkspaceStorageStatus>(
      this.http,
      "GET",
      `/api/v1/workspaces/${this.workspaceId}/storage`,
    );
  }

  listStorageObjects(prefix = "") {
    const query = prefix ? `?prefix=${encodeURIComponent(prefix)}` : "";
    return requestJson<{ items: StorageObjectItem[]; prefix: string }>(
      this.http,
      "GET",
      `/api/v1/workspaces/${this.workspaceId}/storage/objects${query}`,
    );
  }

  uploadStorageObject(file: Blob, options?: { key?: string; fileName?: string }) {
    const form = new FormData();
    form.append("file", file, options?.fileName ?? options?.key ?? "upload");
    const query = options?.key ? `?key=${encodeURIComponent(options.key)}` : "";
    return requestJson<{ key: string; name: string }>(
      this.http,
      "POST",
      `/api/v1/workspaces/${this.workspaceId}/storage/objects${query}`,
      form,
    );
  }

  deleteStorageObject(key: string) {
    return requestJson<void>(
      this.http,
      "DELETE",
      `/api/v1/workspaces/${this.workspaceId}/storage/objects?key=${encodeURIComponent(key)}`,
    );
  }
}

File: sdk/python/dtorch/platform_client.py
"""Project-credential client for workspace platform services."""

from __future__ import annotations

from typing import Any, Dict, List, Optional
from urllib.parse import quote, urlencode

from dtorch.database import DatabaseContext
from dtorch.errors import DtorchValidationError
from dtorch.runtime_client import DtorchRuntimeClient
from dtorch.transport import HttpTransport


class DtorchPlatformClient:
    """Access project-scoped services with a Studio project key and secret."""

    def __init__(
        self,
        base_url: str,
        *,
        project_key: str,
        project_secret: str,
        workspace_id: int,
        database_id: Optional[int] = None,
        timeout: float = 60.0,
    ) -> None:
        self.workspace_id = workspace_id
        self.default_database_id = database_id
        self._http = HttpTransport(
            base_url,
            project_key=project_key,
            project_secret=project_secret,
            timeout=timeout,
        )
        self.runtime = DtorchRuntimeClient(
            base_url,
            workspace_id=workspace_id,
            project_key=project_key,
            project_secret=project_secret,
            timeout=timeout,
        )

    def validate(self) -> Dict[str, Any]:
        response = self.list_databases()
        return {
            "ok": True,
            "workspace_id": self.workspace_id,
            "databases": response.get("databases", []),
        }

    def database(self, database_id: int) -> DatabaseContext:
        return DatabaseContext(self, database_id)

    @property
    def db(self) -> DatabaseContext:
        if self.default_database_id is None:
            raise DtorchValidationError(
                "No default database configured; pass database_id or call database(id)"
            )
        return self.database(self.default_database_id)

    def list_databases(self) -> Dict[str, Any]:
        return self._http.request(
            "GET",
            f"/api/v1/workspaces/{self.workspace_id}/databases",
        )

    def list_tables(self, database_id: int) -> Dict[str, Any]:
        return self._http.request(
            "GET",
            f"/api/v1/workspaces/{self.workspace_id}/databases/{database_id}/tables",
        )

    def get_table_detail(self, database_id: int, table_name: str) -> Dict[str, Any]:
        encoded_name = quote(table_name, safe="")
        return self._http.request(
            "GET",
            f"/api/v1/workspaces/{self.workspace_id}/databases/{database_id}"
            f"/tables/{encoded_name}",
        )

    def get_table_data(
        self,
        database_id: int,
        table_name: str,
        *,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> Dict[str, Any]:
        query_values: Dict[str, int] = {}
        if limit is not None:
            query_values["limit"] = limit
        if offset is not None:
            query_values["offset"] = offset
        query = f"?{urlencode(query_values)}" if query_values else ""
        encoded_name = quote(table_name, safe="")
        return self._http.request(
            "GET",
            f"/api/v1/workspaces/{self.workspace_id}/databases/{database_id}"
            f"/tables/{encoded_name}/data{query}",
        )

    def execute_sql(self, database_id: int, sql: str) -> Dict[str, Any]:
        return self._http.request(
            "POST",
            f"/api/v1/workspaces/{self.workspace_id}/databases/{database_id}/sql",
            {"sql": sql},
        )

    def list_database_migrations(self, database_id: int) -> Dict[str, Any]:
        return self._http.request(
            "GET",
            f"/api/v1/workspaces/{self.workspace_id}/databases/{database_id}/migrations",
        )

    def apply_database_migrations(
        self,
        database_id: int,
        migrations: List[Dict[str, str]],
        *,
        dry_run: bool = False,
    ) -> Dict[str, Any]:
        return self._http.request(
            "POST",
            f"/api/v1/workspaces/{self.workspace_id}/databases/{database_id}/migrations/apply",
            {"migrations": migrations, "dry_run": dry_run},
        )


# Backward-compatible class name.
EltPlatformClient = DtorchPlatformClient
